### ISSUE TITLE
Displaying a textured mesh without vertex colouring

### ISSUE THREAD

**Issue by patrikhuber** [libigl/libigl#1090]
Hi!

When I set some vertices of a mesh in the viewer...:

```
viewer.data().set_vertices(V);
```
and then also uv coordinates and a texture...:
```
viewer.data().set_uv(UV_V, UV_F);
viewer.data().set_texture(R, G, B);
```
then the resulting visualised mesh in the viewer shows the texture but the texture colour is highly tinted yellow. If you display the mesh without texture, then it is indeed yellow (golden), and this golden colour is used by default for meshes. But now when you display a mesh with texture, that golden per-vertex colour is mixed with the texture. I believe that if a texture is set and `show_texture` is `true`, then the vertex colour shouldn't be displayed.

Is there any way to deactivate vertex colouring and just show the texture, with lighting?

I poked around the code quite a bit and I could find that by default, the ambient, diffuse and specular colours are set to `igl::GOLD_*`. So I tried the following:

```
viewer.data().uniform_colors(
                Eigen::Vector3d(igl::WHITE_AMBIENT[0], igl::WHITE_AMBIENT[1], igl::WHITE_AMBIENT[2]),
                Eigen::Vector3d(igl::WHITE_DIFFUSE[0], igl::WHITE_DIFFUSE[1], igl::WHITE_DIFFUSE[2]),
                Eigen::Vector3d(igl::WHITE_SPECULAR[0], igl::WHITE_SPECULAR[1], igl::WHITE_SPECULAR[2]));
```
but the result is even worse, now the texture has loads of very bright white spots. I also can't seem to find any other flags that would change the behaviour of texture/vertex-colour rendering.

(How) is it possible to display the texture, without any colour tinting by the (default or otherwise) vertex colours?

Thanks very much!

**Comment by jdumas** [comment]
You can set the `core().lighting_factor` to `0`, and set the vertices colors to white (using `data().set_colors()`), in which case no lighting is performed and the texture color is mixed with white. If you want to use texture colors as the diffuse color for the mesh ... you'll have to modify the default shader (see [this thread](https://github.com/libigl/libigl/issues/657) for an example on how to do it).

I'm planning to update those shaders at some point, but it's probably a couple of months on the roadmap.

**Comment by patrikhuber** [comment]
Thanks for the hints!

I've already tried setting `core().lighting_factor` to `0` actually. The line you suggested, `data().set_colors(Eigen::RowVector3d(1, 1, 1))`, actually sets the colour to parula (`igl::parula(C,true,C3);` gets called), and the resulting mesh (with and without texture) is still yellow/golden. The line I wrote above with `viewer.data().uniform_colors(...)` actually does set the vertex colour to white (it looks like that at least), but with or without `core().lighting_factor` set to `0`, the resulting visualisation is unfortunately unusable (it is way too white). That's why I was thinking that perhaps I am doing something wrong, because I can't imagine how people would actually use the result of that texturing? Maybe I am indeed still doing something wrong?

**Comment by jdumas** [comment]
The `RowVector` is important here. If you pass a column vector, the method will assume one scalar attribute per vertex, and convert it to a color using the parula colormap. If you pass a matrix with a single row and 3 columns, it will assume the information is a color value, and fill the vertices values directly. `uniform_colors` can be used too, but you need to set the ambient/specular colors accordingly. For example, in the `set_colors()` methods, the ambient and specular terms are deduced as `0.1*diffuse` and `0.3 + 0.1*diffuse` respectively. One can argue about the merits of doing it this way, but it should give you an idea.

That being said, it wouldn't be hard to modify the shader (`mesh_fragment_shader_string` in `MeshGL.cpp`) to mix the diffuse color term and the texture color before computing the shading on it. But you can already customize this part of the pipeline without editing libigl itself (see link above), so it's more like an enhancement than a real bug.

**Comment by patrikhuber** [comment]
Ah, thanks a lot! I removed the `Row` from `RowVector` because I thought I got a compile error and then thought it's some custom type alias from you guys that I didn't have, but it is indeed a type alias from `Eigen` and compiles just fine, not sure what happened :-).

I see! Actually `igl::WHITE_AMBIENT`, `igl::WHITE_DIFFUSE` and `igl::WHITE_SPECULAR` are all just set to the same colour (full white). Without checking it, I would've expected that these already contain those factors for the ambient and specular terms that you mention, because if they were just plain white, there's no reason to have three variables with different values and it could be named just `igl::WHITE`. Anyway my assumption was wrong.
It works now with the `RowVector` and it looks quite okay, thank you!

And thanks also again for the other link - indeed modifying the shader would of course work and would also not be that hard, but I wanted to avoid that because I just wanted to "quickly" display an animated mesh with texture without further effort, and thought that I could do it with the libigl viewer out-of-the-box.

I would be happy to close this, if you think that it is like it is "by design", or leave it open until you get to your plan of updating the included shaders for better texturing support out-of-the-box - As you wish!

Thank you again :)
